### PR TITLE
Introduce execution time value for `UnionFileCollection`

### PR DESCRIPTION
**Pull request by 6hundreds**
Part of https://github.com/gradle/gradle/issues/28820

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bot-gradle test this

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java [hunk 1]
```diff
@@ -81,4 +84,35 @@ protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
...
+
+        public UnionExecutionTimeValue(ImmutableSet<FileCollectionExecutionTimeValue> source) {
+            this.source = source;
+        }
+
+        @Override
+        public FileCollectionInternal toFileCollection(FileCollectionFactory fileCollectionFactory) {
+            Set<FileCollectionInternal> source = this.source.stream()
+                .map(value -> value.toFileCollection(fileCollectionFactory))
+                .collect(Collectors.toSet());
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java:113]: ❓ Why do it in two steps?

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java [hunk 2]
```diff
@@ -81,4 +84,35 @@ protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
             visitor.accept(fileCollection);
         }
     }
+
+    @Override
+    public Optional<FileCollectionExecutionTimeValue> calculateExecutionTimeValue() {
+        Set<FileCollectionExecutionTimeValue> sourceExecutionTimeValues = new HashSet<>();
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java:90]: ❌ Use an `ImmutableSet.Builder` for efficiency and correctness (order preserving)

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java (hunk 1 above)
Comment (6hundreds) [reply platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java:113]: Did you mean something like
```
Set<FileCollectionInternal> source = this.source.stream()
                .collect(Collectors.mapping(
                    executionTimeValue -> executionTimeValue.toFileCollection(fileCollectionFactory),
                    Collectors.toSet()
                ))
```
?
If so, there is an actionable warning
Screenshot 2024-07-22 at 17 22 02

**Comment by 6hundreds** [comment]
@bamboo ready to final look

File: platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            fileCollectionFactory().fixed("foo", new File("foo.txt")),
+            fileCollectionFactory().fixed("bar", new File("bar.txt"))
+        ]
+        def collection = new UnionFileCollection(DefaultTaskDependencyFactory.withNoAssociatedProject(), source)
+
+        when:
+        def executionTimeValue = collection.calculateExecutionTimeValue()
+
+        then:
+        executionTimeValue.present
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy:38]: 🧪 We're missing test coverage for the case of an absent collection

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java [hunk 4]
```diff
@@ -81,4 +82,33 @@ protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
...
+    private static class UnionExecutionTimeValue implements FileCollectionExecutionTimeValue {
+        private final ImmutableSet<FileCollectionExecutionTimeValue> source;
+
+        public UnionExecutionTimeValue(ImmutableSet<FileCollectionExecutionTimeValue> source) {
+            this.source = source;
+        }
+
+        @Override
+        public FileCollectionInternal toFileCollection(FileCollectionFactory fileCollectionFactory) {
+            return source.stream()
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java:109]: 👍

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy (hunk 3 above)
Comment (6hundreds) [reply platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy:38]: @bamboo under `absent` you meant "file collection with no execution time value?"

Comment (bamboo) [reply platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy:38]: @6hundreds right, I got the terminology mixed up with providers. An empty execution time value is what I mean

File: platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/EmptyExecutionTimeValueFileCollection.groovy [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file
+
+class EmptyExecutionTimeValueFileCollection extends AbstractFileCollection{
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/EmptyExecutionTimeValueFileCollection.groovy:19]: 💅 Format file

File: platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        def newCollection = executionTimeValue.get().toFileCollection(fileCollectionFactory())
+
+        then:
+        newCollection.files ==  [new File("foo.txt"), new File("bar.txt")] as Set
+        newCollection.displayName == "file collection"
+    }
+
+    def "UnionFileCollection supports absence of execution time value for source"() {
+        given:
+        def emptyEtvSource = new EmptyExecutionTimeValueFileCollection()
```
Comment (bamboo) [inline platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/UnionFileCollectionExecutionTimeValueTest.groovy:50]: 💅 Avoid non-pronounceable names, `emptySource` is fine.

**Review by bamboo — APPROVED** [review]
Left a couple of requests, other than those it looks good 👍

**Pull request merged**